ddclient.in

```diff
@@ -7940,6 +7970,469 @@ sub nic_infomaniak_update {
+    my $request_url = $params{'request_url'};
+    my $request_method = $params{'request_method'} // "GET";
+    my $request_payload = $params{'request_payload'} // "";
+    my $request_headers = $params{'request_headers'} // {};
+
+    # Error will bubble up to primary update function done to prevent extra logic in
+    # main function required to parse different return types
+    my $url = parse_url($request_url);
+    my $date = create_date();
+
+    my $canonical_request = create_canonical_request_hash(
+        $request_method,
+        $url,
+        $request_payload,
+        $request_headers,
+        $date
+    );
+
+    # Interesting so when pulling out by key if not within another variable the keys is pulled too....
+    my $canonical_request_hash = %$canonical_request{hash};
+    my $string_to_sign = create_string_to_sign($canonical_request_hash, $service, $region, $date);
+
+    my $result_string = %$string_to_sign{string};
+    my $signature = create_signature($result_string, $aws_secret_access_key, $region, $service, $date);
+
+    # curl subroutine above is expected headers to be a string
+    my $aws_authorization = $ALGORITHM." Credential=".$aws_access_key_id."/".%$string_to_sign{"scope"}.",SignedHeaders=".%$canonical_request{"signed_headers"}.",Signature=".$signature;
+
+    debug("\nAuthorization:".$aws_authorization."\n\n");
+    my $request_header_string = "";
+    my %compiled_headers = (%{%$canonical_request{"finalized_headers"}}, ( Authorization => $aws_authorization));
+    my @compiled_headers_key = sort keys %compiled_headers;
+    for my $header_key (@compiled_headers_key) {
+        $request_header_string = $request_header_string.(lc $header_key).":".($compiled_headers{$header_key});
+
+        if ($header_key ne $compiled_headers_key[$#compiled_headers_key]) {
+            $request_header_string = $request_header_string."\n";
+        }
+    }
+
+    my %curl_opts = (
+        url => get_encoded_url($url),
+        method => $request_method,
+        headers => $request_header_string
+    );
+
+    if ($request_method =~ "PUT|POST|PATCH") {
+        debug("Body: ".$request_payload."\n");
+        $curl_opts{data} = $request_payload;
+    }
+
+    return geturl(%curl_opts);
+}
+
+my $AWS_ROUTE53_BASE = "https://route53.amazonaws.com/2013-04-01/";
+sub get_route53_api_url {
+    my (
+        $resource,
+        $query_parameters
+    ) = @_;
+
+    my $api_request_url = $AWS_ROUTE53_BASE.$resource;
+
+    if (defined($query_parameters)) {
+        $api_request_url .= "?".$query_parameters
+    }
+
+    return $api_request_url;
+}
+
+sub update_route53_one {
+    my (
+        $hosted_zone_id,
+        $h,
+        $resource_set_type,
+        $ip,
+        $ttl,
+        $aws_access_key_id,
+        $aws_secret_access_key,
+        $aws_region
+    ) = @_;
+
+    my $resource_type = $resource_set_type // ($ip =~ /\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}/ ? "A" : "AAAA");
```

Comment:
There already are helper functions called `is_ipv4` and `is_ipv6` for this.
Though in this case you could make this easier by having `update_route53_one` not as a separate function but as part of `nic_route53_update`. Then you can do this fairly common pattern from other providers:

```perl
sub nic_route53_update {
    debug("\nnic_route53_update---------------------");

    foreach my $h (@_) {
        my $ipv4 = delete $config{$h}{'wantipv4'};
        my $ipv6 = delete $config{$h}{'wantipv6'};
        verbose("UPDATE:", "updating %s", $h);

        foreach my $ip ($ipv4, $ipv6){
            next if (!$ip);
            info("setting IP address to %s for %s", $ip, $h);
            my $ipv  = ($ip eq ($ipv6 // '')) ? '6' : '4';
            my $type = ($ip eq ($ipv6 // '')) ? 'AAAA' : 'A';

            # Logic from update_route53_one
            # ...
        }
    }
}
```

Reply:
Adopted, much cleaner, thank you for the advice!